### PR TITLE
Main storage adapter abstraction + postgre adapter

### PR DESCRIPTION
**Pull request by henry-hz**
Dear @slashdotdash , once again, thanks for sharing Commanded. This pull request contains the main structure for loading and appending state and events using a multi-db adapter architecture. If it's fine, I can continue to implement the connection start/stop, support for multi-db in all tests, event subscribing, and add extreme (greg db). Note that I made some polish in names and docs to optimize the view for smaller monitors (1920x1080).. for me it's important to reason about the module as whole while coding.
I followed the idea of encapsulating all the specific db code into the adapter folder, so specific mappers also are moved there, and Commanded deals only with raw event data and state.

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -38,7 +38,12 @@ defmodule Commanded.Mixfile do
       {:markdown, github: "devinus/markdown", only: :dev},
       {:mix_test_watch, "~> 0.2", only: :dev},
       {:poison, "~> 3.0"},
-      {:uuid, "~> 1.1"}
+      {:uuid, "~> 1.1"},
+      # devs
+      {:dialyxir, "~> 0.3.5", only: :dev},   # simplify dialyzer, type: mix dialyzer.plt first
+			{:mix_test_watch, "~> 0.2", only: :dev} # use mix test.watch for adicted TDD development
```
Comment (slashdotdash) [inline mix.exs]: `mix_test_watch` is already included as a dependency (line 39).

File: lib/commanded/storage/storage.ex [hunk 2]
```diff
@@ -0,0 +1,64 @@
+defmodule Commanded.Storage.Storage do
```
Comment (henry-hz) [inline lib/commanded/storage/storage.ex]: sure

File: mix.exs (hunk 1 above)
Comment (henry-hz) [inline mix.exs]: ok!

**Comment by henry-hz** [comment]
It's okay with the idea of Eventstore and Extreme, nice to have more SQL around. Should I make the changes and pull request or you merge and I fork from you again?
thanks

**Comment by henry-hz** [comment]
Notice that travis sometimes fail.

**Comment by slashdotdash** [comment]
[GitHub flow](https://guides.github.com/introduction/flow/) is a simpler approach. There's only one rule: anything in the `master` branch is always deployable.  

So each new features has its own branch (e.g. `feature/storage-abstraction`) that get merged into `master` once it is *done*. I consider that to mean it is fully unit/integration tested, has no commented out or redundant code, and is documented (in-line docs, README).

We could make this PR the initial storage abstraction with the single adapter using the PostgreSQL `eventstore`. Then a subsequent PR containing the Extreme adapter.

**Comment by henry-hz** [comment]
Ok, Github flow is a good approach. How should I proceed ? I understand that you will create a branch to hold this PR ? Because process_manager_router subscribes to all events, I am still figuring out how to do it in Extreme... isn't too much ? maybe better every process_manager_instance subscribe to it's interested events ?

**Comment by slashdotdash** [comment]
I don't think GitHub allows you to change the target branch of a pull requests. You will need to close this one and open a new pull request to target another branch. You may even need to branch in your local git repo first and PR from there.

**Comment by henry-hz** [comment]
ok!

**Pull request closed without merging**